presto-main/src/main/java/com/facebook/presto/operator/PartitionedLookupSourceFactory.java

```diff
@@ -101,52 +142,221 @@ public PartitionedLookupSourceFactory(List<Type> types, List<Type> outputTypes,
 
     public void setPartitionLookupSourceSupplier(int partitionIndex, Supplier<LookupSource> partitionLookupSource)
     {
-        requireNonNull(partitionLookupSource, "partitionLookupSource is null");
+        SetLookupSourceResult result = internalSetLookupSource(partitionIndex, partitionLookupSource);
 
-        Supplier<LookupSource> lookupSourceSupplier = null;
-        List<SettableFuture<LookupSource>> lookupSourceFutures = null;
-        synchronized (this) {
-            if (destroyed.isDone()) {
-                return;
+        if (result.getLookupSourceSupplier() != null) {
+            for (SettableFuture<LookupSource> lookupSourceFuture : lookupSourceFutures) {
+                lookupSourceFuture.set(lookupSourceSupplier.get());
             }
+        }
+    }
 
-            checkState(partitions[partitionIndex] == null, "Partition already set");
-            partitions[partitionIndex] = partitionLookupSource;
-            partitionsSet++;
+    public void setPartitionSpilledLookupSourceSupplier(int partitionIndex, SingleStreamSpiller lookupSourceSpiller)
+    {
+        SetLookupSourceResult result;
+        synchronized (this) {
+            requireNonNull(lookupSourceSpiller, "lookupSource is null");
 
-            if (partitionsSet == partitions.length) {
-                if (partitionsSet != 1) {
-                    List<Supplier<LookupSource>> partitions = ImmutableList.copyOf(this.partitions);
-                    this.lookupSourceSupplier = createPartitionedLookupSourceSupplier(partitions, hashChannelTypes, outer);
-                }
-                else if (outer) {
-                    this.lookupSourceSupplier = createOuterLookupSourceSupplier(partitionLookupSource);
-                }
-                else {
-                    this.lookupSourceSupplier = partitionLookupSource;
-                }
+            spilledLookupSources.put(partitionIndex, lookupSourceSpiller);
 
-                // store lookup source supplier and futures into local variables so they can be used outside of the lock
-                lookupSourceSupplier = this.lookupSourceSupplier;
-                lookupSourceFutures = ImmutableList.copyOf(this.lookupSourceFutures);
-            }
+            result = internalSetLookupSource(partitionIndex, () -> new SpilledLookupSource(types.size()));
```

Comment:
What's the point of storing SpilledLookupSource-s in a `Supplier<LookupSource>[]`?  Wouldn't `LookupSource[]` be enough? Or can the `types` change in the meantime and we're working around that?